system-addon/lib/DownloadsManager.jsm

```diff
@@ -0,0 +1,146 @@
+ChromeUtils.import("resource://gre/modules/Services.jsm");
+Cu.importGlobalProperties(["URL"]);
+
+const {actionTypes: at} = ChromeUtils.import("resource://activity-stream/common/Actions.jsm", {});
+
+ChromeUtils.defineModuleGetter(this, "DownloadsViewUI",
+  "resource:///modules/DownloadsViewUI.jsm");
+ChromeUtils.defineModuleGetter(this, "DownloadsCommon",
+  "resource:///modules/DownloadsCommon.jsm");
+
+const RECENT_DOWNLOAD_THRESHOLD = 36 * 60 * 60 * 1000;
+
+class DownloadElement extends DownloadsViewUI.DownloadElementShell {
+  constructor(download, browser) {
+    super();
+    this._download = download;
+    this.element = browser;
+    this.element._shell = this;
+  }
+
+  get download() {
+    return this._download;
+  }
+
+  get downloadedSince() {
+    return Date.now() - this.download.endTime;
+  }
+
+  get fileType() {
+    if (!this.download.target.path) {
+      return null;
+    }
+    let items = this.download.target.path.split(".");
+    return items[items.length - 1].toUpperCase();
+  }
+
+  downloadsCmd_copyLocation() {
+    let clipboard = Cc["@mozilla.org/widget/clipboardhelper;1"].getService(Ci.nsIClipboardHelper);
+    clipboard.copyString(this.download.source.url);
+  }
+}
+
+this.DownloadsManager = class DownloadsManager {
+  constructor(store) {
+    this._downloadData = null;
+    this._window = null;
+    this._store = null;
+    this._viewableDownloadItems = new Map();
+  }
+
+  formatDownload(element) {
+    const downloadedItem = element.download;
+    let description;
+    if (element.fileType) {
+      // If we have a file type: '1.5 MB — PNG'
+      description = `${element.sizeStrings.stateLabel} \u2014 ${element.fileType}`;
+    } else {
+      // If we do not have a file type: '1.5 MB'
+      description = `${element.sizeStrings.stateLabel}`;
+    }
+    return {
+      hostname: new URL(downloadedItem.source.url).hostname,
+      url: downloadedItem.source.url,
+      path: downloadedItem.target.path,
+      title: element.displayName,
+      description,
+      referrer: downloadedItem.source.referrer,
+      date_added: downloadedItem.endTime
+    };
+  }
+
+  init(store) {
+    this._store = store;
+    this._browser = Services.appShell.hiddenDOMWindow;
+    this._downloadData = DownloadsCommon.getData(this._browser.ownerGlobal, true, false, true);
+    this._downloadData.addView(this);
+  }
+
+  onDownloadAdded(download) {
+    const elem = new DownloadElement(download, this._browser);
+    const downloadedItem = elem.download;
+    if (!this._viewableDownloadItems.has(downloadedItem.source.url)) {
+      this._viewableDownloadItems.set(downloadedItem.source.url, elem);
+      this._store.dispatch({type: at.DOWNLOAD_CHANGED});
```

Comment:
Looks like this is called when it first starts a download or an existing download. We probably want handle downloads that were just started and then completed to get the correct file size, otherwise:
<img width="343" alt="image" src="https://user-images.githubusercontent.com/438537/38959313-fff80b56-4314-11e8-92e9-e2450a2b8625.png">

Dismissing some page forces a refresh and it correctly shows:
<img width="340" alt="image" src="https://user-images.githubusercontent.com/438537/38959337-134545ca-4315-11e8-99f0-c6005d45a438.png">